generator/generate.py

```diff
@@ -687,31 +686,56 @@ def parse_groups(self, match_t, match_re, ends=';'):
                     d = []
         f.close()
 
-    def parse_param(self, param):
+    def parse_param(self, param_raw):
         """Parse a C parameter expression.
 
         It is used to parse the type/name of functions
         and type/name of the function parameters.
 
         @return: a Par instance.
         """
-        t = param.replace('const', '').strip()
-        if _VLC_FORWARD_ in t:
-            m = forward_re.match(t)
-            t = m.group(1) + m.group(2)
-
-        m = param_re.search(t)
-        if m:
-            _, t, n = m.groups()
-            while n.startswith('*'):
-                n  = n[1:].lstrip()
-                t += '*'
-##          if n == 'const*':
-##              # K&R: [const] char* const*
-##              n = ''
-        else:  # K&R: only [const] type
-            n = ''
-        return Par(n, t.replace(' ', ''))
+        param_raw = param_raw.strip()
+        if _VLC_FORWARD_ in param_raw:
+            m = forward_re.match(param_raw)
+            param_raw = m.group(1) + m.group(2)
+
+        # is this parameter a pointer?
+        split_pointer = param_raw.split('*')
+        if len(split_pointer) > 1:
+            param_name = split_pointer[-1]
+            # TODO extract constness
+            param_type = split_pointer[0].replace('const', '').strip()
+            # remove the struct keyword, this information is currently not used
+            param_type = param_type.replace('struct ', '').strip()
+
+            param_type += '*' * (len(split_pointer) - 1 )
+        # ... or is it a simple variable?
+        else:
+            # WARNING: workaround for "union { struct {"
+            param_raw = param_raw.split('{')[-1]
+
+            # ASSUMPTIONs
+            # these allows to constrain param_raw to these options:
+            #  - named:     "type name" (e.g. "int param")
+            #  - anonymous: "type"      (e.g. "int")
+            assert('struct' not in param_raw)
+            assert('const' not in param_raw)
+
+            # normalize spaces
+            param_raw = re.sub('\s+', ' ', param_raw)
+            # TODO remove struct and const
+            try:
+                split_value = param_raw.split(' ')
+                if len(split_value) > 1:
+                    param_name = split_value[-1]
+                    param_type = ' '.join(split_value[:-1])
+                else:
+                    param_type = split_value[0]
+                    param_name = ''
+            except:
+                param_name = ''
```

Comment:
Shouldn't we initialize param_type also here? And specify IndexError for the exception, catching full exceptions can lead to unexpected errors.